src/link/inaccessible_spec_definition.rs

```diff
@@ -0,0 +1,1137 @@
+    ) -> Result<Directive, FederationError> {
+        let name_in_schema = self
+            .directive_name_in_schema(schema, &INACCESSIBLE_DIRECTIVE_NAME_IN_SPEC)?
+            .ok_or_else(|| SingleFederationError::Internal {
+                message: "Unexpectedly could not find inaccessible spec in schema".to_owned(),
+            })?;
+        Ok(Directive {
+            name: name_in_schema,
+            arguments: Vec::new(),
+        })
+    }
+}
+
+impl SpecDefinition for InaccessibleSpecDefinition {
+    fn url(&self) -> &Url {
+        &self.url
+    }
+
+    fn minimum_federation_version(&self) -> Option<&Version> {
+        self.minimum_federation_version.as_ref()
+    }
+}
+
+lazy_static! {
+    pub(crate) static ref INACCESSIBLE_VERSIONS: SpecDefinitions<InaccessibleSpecDefinition> = {
+        let mut definitions = SpecDefinitions::new(Identity::inaccessible_identity());
+        definitions.add(InaccessibleSpecDefinition::new(
+            Version { major: 0, minor: 1 },
+            None,
+        ));
+        definitions.add(InaccessibleSpecDefinition::new(
+            Version { major: 0, minor: 2 },
+            Some(Version { major: 2, minor: 0 }),
+        ));
+        definitions
+    };
+}
+
+fn get_inaccessible_spec_from_schema(
+    schema: &FederationSchema,
+) -> Result<Option<&'static InaccessibleSpecDefinition>, FederationError> {
+    let inaccessible_link = match schema
+        .metadata()
+        .as_ref()
+        .and_then(|metadata| metadata.for_identity(&Identity::inaccessible_identity()))
+    {
+        None => return Ok(None),
+        Some(link) => link,
+    };
+    Ok(Some(
+        INACCESSIBLE_VERSIONS
+            .find(&inaccessible_link.url.version)
+            .ok_or_else(|| SingleFederationError::Internal {
+                message: format!("Cannot remove inaccessible elements: the schema uses unsupported inaccessible spec version {}", inaccessible_link.url.version),
+            })?,
+    ))
+}
+
+fn is_type_system_location(location: DirectiveLocation) -> bool {
+    matches!(
+        location,
+        DirectiveLocation::Schema
+            | DirectiveLocation::Scalar
+            | DirectiveLocation::Object
+            | DirectiveLocation::FieldDefinition
+            | DirectiveLocation::ArgumentDefinition
+            | DirectiveLocation::Interface
+            | DirectiveLocation::Union
+            | DirectiveLocation::Enum
+            | DirectiveLocation::EnumValue
+            | DirectiveLocation::InputObject
+            | DirectiveLocation::InputFieldDefinition
+    )
+}
+
+fn field_uses_inaccessible(field: &FieldDefinition, inaccessible_directive: &Name) -> bool {
+    field.directives.has(inaccessible_directive)
+        || field
+            .arguments
+            .iter()
+            .any(|argument| argument.directives.has(inaccessible_directive))
+}
+
+/// Check if a type definition uses the @inaccessible directive anywhere in its definition.
+fn type_uses_inaccessible(
```

Comment:
since this is only used for validating core feature types should we rename it to more specific?

Reply:
it's currently only used for core feature types but the functionality isn't specific to core feature types.